### PR TITLE
UI: Design a 404 page for default redirects

### LINKED ISSUE #33: UI: Design a 404 page for default redirects
### Is your feature request related to a problem? Please describe.
Currently, wrong URLs redirect to a general page but the correct redirect should be to a `404: Not found` page. We need the design of such a page and a change in the URL `Routes.js`

### Describe the solution you'd like
- Design a 404 page
- Change route URL to Redirect to `/404` on wrong URLs

### Definition of Done
- [ ] A 404 error page mockup
- [ ] A change in `Routes.js` for proper redirect

### PR DESCRIPTION
**Pull request by Aaishpra**
### Description
Solves the Following.

   

-  Design a 404 page

   

-  Change route URL to Redirect to /404 on wrong URLs

Fixes #33

### Type of Change:

- Code

- User Interface

**Code/Quality Assurance Only**

- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?

ScreenShot for Reference
image

### Checklist:

- [x]  My PR follows the style guidelines of this project

- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] Any dependent changes have been merged
- [x] I have written Kotlin Docs whenever is applicable

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings
- [x] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been published in downstream modules

**Additional Context**
 I initially wanted to add a button to go back to the home page but since it doesn't Exist hence skipped that part. More suggestions and Additions are welcome for any other Feature within the page

### CODE REVIEW

**Comment by bismitaguha** [comment]
You should have used AuthRoutes IMO!

**Comment by Aaishpra** [comment]
Done @bismitaguha please review

File: src/components/ErrorPage.js
Comment (bismitaguha) [inline src/components/ErrorPage.js]: AuthRoute already contains the Navbar, so this wasn't needed. Reusing components is always preferable. That's why I suggested AuthRoute.

File: src/components/ErrorPage.js
Comment (bismitaguha) [inline src/components/ErrorPage.js:40]: Throughout the application class method has been used.

**Review by bismitaguha — CHANGES_REQUESTED** [review]

**Comment by Aaishpra** [comment]
Changed the files @bismitaguha

**Comment by Aaishpra** [comment]
@bismitaguha please review the changes

**Comment by Aaishpra** [comment]
@abha224 can u review the pull request as @bismitaguha seems not to be available

File: src/components/ErrorPage.js
Comment (bismitaguha) [reply src/components/ErrorPage.js:40]: @Aaishpra Make these changes

**Comment by bismitaguha** [comment]
I had requested changes @Aaishpra .Please make them!

File: src/components/ErrorPage.js
Comment (Aaishpra) [reply src/components/ErrorPage.js:40]: Already done @bismitaguha

**Comment by Aaishpra** [comment]
@bismitaguha can u please check my latest commits i guess i changed the file as u requested.

**Comment by bismitaguha** [comment]
@Aaishpra I didn't ask you to remove classes. The code already written is in exporting a class component not by exporting default function methods
Just go through the `src/` directory once.

**Comment by sidvenu** [comment]
@Aaishpra is the 404 text color the same as the color of Open Source Programs header? If not, I suggest switching to that color.

**Comment by Aaishpra** [comment]
@bismitaguha does it needs to be done by using class ErrorBoundary 
@sidvenu the color is the same

File: src/components/ErrorPage.js
Comment (Aaishpra) [reply src/components/ErrorPage.js:40]: @bismitaguha is this correct now

**Comment by Aaishpra** [comment]
@sidvenu the color is the same maybe  it looks a bit different because the text 404 is much bolder than Open source programs header.

**Comment by Aaishpra** [comment]
@bismitaguha is the latest commit ok now?

**Comment by Aaishpra** [comment]
@bismitaguha please review the changes

**Comment by bismitaguha** [comment]
@Aaishpra Either me or someone else will surely review your PR, from time to time. Considering my situation, I am running busy lately, and so are many other members of the community too. Please have patience and don't worry at all. Your PRs will surely be reviewed. Although, if a week or more passes surely ping/mention any of us on Github/Zulip.

File: src/components/ErrorPage.js
Comment (bismitaguha) [reply src/components/ErrorPage.js:40]: Yes

**Review by bismitaguha — DISMISSED** [review]

**Comment by Aaishpra** [comment]
No Problem @bismitaguha got your point.

**Comment by Aaishpra** [comment]
@sidvenu can you please review this.

**Review by sidvenu — DISMISSED** [review]
LGTM!

**Comment by sidvenu** [comment]
@anitab-org/qa-team @anitab-org/open-source-progs-maintainers  @anitab-org/open-source-progs-developers Can anyone test this PR out?

**Comment by Aaishpra** [comment]
Hey,@bismitaguha and @sidvenu can you suggest someone that can test this PR out so that i can personally tag them here, or you people can help with it. Actually a lot of time has already passed since the last review, and no one has come to see this here.

**Comment by sidvenu** [comment]
@rpattath @bismitaguha if possible, could you test this PR?

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
@Aaishpra please squash your commits into one. It's always better to have one commit per change to keep the history clean

File: src/styles/ErrorPage.css
Comment (devkapilbansal) [inline src/styles/ErrorPage.css]: Why two blank lines are here? We only need one

**Comment by Aaishpra** [comment]
@devkapilbansal can you tell me what are the results of testing.
Also @bismitaguha @sidvenu please review this again.

**Comment by devkapilbansal** [comment]
@Aaishpra I haven't tested it yet. At first you have to squash your commits in one following the commit message guidelines. Then I will test it

**Review by Aaishpra — COMMENTED** [review]
@rpattath since, i was having free time i tested my PR by myself (with the help of doc you provided on zulip ). I hope that's Ok. The PR already had two approvals by @bismitaguha and @sidvenu . But, since i had to squash my commits and remove spaces as suggested the approvals were canceled. So, @bismitaguha and @sidvenu please review this again.

The changes made in this PR were tested locally. Following are the results:

1. Code review - Done 

2. All possible responses (positive and negative tests) were tested as below:

   * _Test1 Description_  : Error Page should appear when the urls are wrong or broken
     _Screenshot/gif_: 
image
Gif link:https://drive.google.com/file/d/1V-h68CBC4jeAbMB4QFr_19J1IO-E0Izp/view?usp=sharing

     _Expected Result_:  The error page should appear.
     _Actual Result_: Same as Expected result.
   * _Test2 Description_  : UI should be responsive
     _Screenshot/gif_:  
image
Gif link: https://drive.google.com/file/d/1yHvXsRmFlJSikys8yfcZdoNJYsQkceaT/view?usp=sharing
     _Expected Result_:  UI is fully responsive.
     _Actual Result_:  same as expected result.
     ...  
  

4. Additional Comments: N/A


5. Status of PR Changed to: Needs Review.

**Comment by khushishikhu** [comment]
I have checked this PR. This PR is ready to merge with the approval of maintainer. LGTM @Aaishpra

**Review by vj-codes — COMMENTED** [review]
@Aaishpra this looks nice , here's a suggestion from my end -
It will be more user friendly if a button taking the user to the Home page is added below the error text , what say ? @isabelcosta 
If decided otherwise the PR is good to go:)

**Comment by Aaishpra** [comment]
> @Aaishpra this looks nice , here's a suggestion from my end -
> It will be more user friendly if a button taking the user to the Home page is added below the error text , what say ? @isabelcosta 
> If decided otherwise the PR is good to go:)

There is no homepage yet in the application, wanted to add one but couldn't. 😆

**Comment by vj-codes** [comment]
@Aaishpra ahh okay then 👀

**Review by vj-codes — DISMISSED** [review]
LGTM 🎉

**Comment by vj-codes** [comment]
@devkapilbansal hey can you review this ?

**Review by isabelcosta — DISMISSED** [review]
@Aaishpra great work! Can't wait to see this on the website 🤗 
Thank you for this contribution!

**Comment by vj-codes** [comment]
@isabelcosta this PR is tested above by @Aaishpra and all details are added as well
https://github.com/anitab-org/open-source-programs-web/pull/57#pullrequestreview-516164173

**Comment by isabelcosta** [comment]
> this PR is tested above by @Aaishpra and all details are added as well

@vj-codes I see and I love it! However, when we think of testing PRs, it has to be someone who is not the author of the PR.

**Comment by INNOVATIVEGAMER** [comment]
@Aaishpra You added custom css for all your styling. Styling is promising but this **project is already using Semantic-UI as UI framework**. I added some comments inside your css with some semantic-ui links which you can use in your component.
You can still use your custom one or two line css by adding `style={{}}` in components prop directly.
@isabelcosta I highly recommend you to add guidelines for usage of semantic-ui in designing of this repo.

**Comment by isabelcosta** [comment]
@INNOVATIVEGAMER good point, could you please create a issue for us to add this to the guidelines please?
@Aaishpra could you take a look at this. How feasible it is to use this instead of custom styling? https://semantic-ui.com/

**Comment by INNOVATIVEGAMER** [comment]
> @INNOVATIVEGAMER good point, could you please create a issue for us to add this to the guidelines please?

I created the issue for this #65. You can check on that. 
Also the contributing guidelines link is not working in readme with some mistakes, I created the issue for same also in #64 .

**Review by devkapilbansal — COMMENTED** [review]
@vj-codes @isabelcosta I am not sure about coding practices and all.
But when I tested it locally, the UI changes looks good to me

The changes made in this PR is tested locally, following are the results:-
1) Code Review - *Not Done*

2) All possible responses were tested as below:
 
- Test1 Description - 404 error page appears when urls are wrong
Expected Result:-  The error page should appear
Actual Result:- Same as expected

- Test2 Description - The UI should be responsive
Actual Result same as expected
Open Source Programs

**Comment by isabelcosta** [comment]
@devkapilbansal that is just awesome! thank you so much!

**Comment by Aaishpra** [comment]
> @Aaishpra You added custom css for all your styling. Styling is promising but this **project is already using Semantic-UI as UI framework**. I added some comments inside your css with some semantic-ui links which you can use in your component.
> You can still use your custom one or two line css by adding `style={{}}` in components prop directly.
> @isabelcosta I highly recommend you to add guidelines for usage of semantic-ui in designing of this repo.

Thanks for the suggestion, I will update soon with the changes.

**Comment by Aaishpra** [comment]
@isabelcosta I have modified the files according to the suggestion, without any change in designs. Here, I would like to mention one thing earlier I tried using `<Container Text>` from semantic-UI but it was breaking the design without centering the content, so I still have used some of my custom CSS to center the content and keep the designs same.

image

Hope this is good.

**Review by codesankalp — COMMENTED** [review]
Tested Locally @isabelcosta @Aaishpra @vj-codes 

Details of the test mentioned below:
**Code-Review: Not Done.**
**Tests:**
1. The 404 error page for invalid URL when not logged in.
     __Screenshot__:
     
image

2. 404 Error page not showing when logged in it redirects to home page maybe this is due to AuthRoute.
__GIF__:
captured gif

3. The scaling works perfect but for smaller width and height I observed this:
__Screenshot__:
image

Also, Please squash your commits @Aaishpra .

**Comment by isabelcosta** [comment]
I can squash so no need to squash @Aaishpra :)
@devkapilbansal and @codesankalp can you please add a code review?
@Aaishpra can you try to get more reviews from the community please, when it has 2 reviews I can merge :)

File: src/styles/ErrorPage.css
Comment (Rahulm2310) [inline src/styles/ErrorPage.css:7]: @Aaishpra Can I ask why are you using absolute positioning when this can be achieved with relative positioning as well. This way the overlapping issue on small screens mentioned by @codesankalp will also resolve.

**Comment by Aaishpra** [comment]
This is what i see on my screen when changing ` position ` from `absolute` to ` relative ` (also the current design is working fine for smaller heights and width as tested by @devkapilbansal and i don't know how to make it work for even smaller widths)
image
about the` AuthRoute` part using of ` AuthRoute ` was suggested to me by @bismitaguha during code review of this PR (and i  also think that it's the correct way to do the redirect instead of using `Route`).

cc : @isabelcosta @codesankalp @Rahulm2310 @devkapilbansal

File: src/styles/ErrorPage.css
Comment (Aaishpra) [reply src/styles/ErrorPage.css:7]: ` relative` is not gonna work on larger screen

**Comment by Rahulm2310** [comment]
@Aaishpra Not just doing `position : relative` would work. Try adding `width :100% height:100vh`. As you already have `display flex, justify-content:center, align-items:center`, I think this should work. 🙂

**Comment by Aaishpra** [comment]
> @Aaishpra Not just doing `position : relative` would work. Try adding `width :100% height:100vh`. As you already have `display flex, justify-content:center, align-items:center`, I think this should work. slightly_smiling_face

Not working

**Comment by khushishikhu** [comment]
I have tested the PR. LGTM!! keep it up @Aaishpra

**Review by khushishikhu — APPROVED** [review]

File: src/styles/ErrorPage.css
Comment (codesankalp) [inline src/styles/ErrorPage.css:26]: Can we use className (err-num and err-text) instead of `style={{}}` for Header in `ErrorPage.js` and add the styles in `ErrorPage.css` as mentioned below?
```suggestion
}

.ui.header.err-num {
    color: #2185D0;
    font-size: 10vw;
    font-weight: bolder;
    line-height: 1em;
    text-transform: uppercase;
}

.ui.header.err-text {
    font-size: 2em;
    margin-bottom: 20px;
    text-transform: uppercase;
    max-width: 600px;
    color: #0d0d0d;
}
```

**Review by codesankalp — APPROVED** [review]
Code review is done. 
LGTM :+1:

File: src/styles/ErrorPage.css
Comment (isabelcosta) [reply src/styles/ErrorPage.css:26]: Any specific reason @codesankalp ? Is this a common convention?

**Comment by isabelcosta** [comment]
Thank you all for the reviews :) I will merge it, and some comments about styling can be done at a later issue.

**Pull request merged**
